processor/src/fast/basic_block.rs

```diff
@@ -165,58 +250,150 @@ impl FastProcessor {
+                return ControlFlow::Break(BreakReason::Err(ExecutionError::advice_error(
+                    err, clk, err_ctx,
+                )));
+            }
+        }
+        ControlFlow::Continue(())
+    }
+
+    /// Execute the finish phase of a basic block node.
+    #[expect(clippy::too_many_arguments)]
+    #[inline(always)]
+    pub(super) fn finish_basic_block(
+        &mut self,
+        basic_block_node: &BasicBlockNode,
+        node_id: MastNodeId,
+        current_forest: &Arc<MastForest>,
+        host: &mut impl AsyncHost,
+        continuation_stack: &mut ContinuationStack,
+        tracer: &mut impl Tracer,
+        stopper: &impl Stopper,
+    ) -> ControlFlow<BreakReason> {
+        tracer.start_clock_cycle(
+            self,
+            NodeExecutionState::End(node_id),
+            continuation_stack,
+            current_forest,
+        );
+
+        // Corresponds to the row inserted for the END operation added to the trace.
+        self.increment_clk(tracer, stopper).map_break(|_| {
+            BreakReason::Stopped(Some(Continuation::AfterExitDecoratorsBasicBlock(node_id)))
+        })?;
+
+        self.execute_end_of_block_decorators(basic_block_node, node_id, current_forest, host)?;
+        self.execute_after_exit_decorators(node_id, current_forest, host)
+    }
+
+    // Executes any decorators which have not been executed during span ops execution; this can
+    // happen for decorators appearing after all operations in a block. these decorators are
+    // executed after BASIC BLOCK is closed to make sure the VM clock cycle advances beyond the last
+    // clock cycle of the BASIC BLOCK ops. For the linked case, check for decorators at an operation
+    // index beyond the last operation
+    #[inline(always)]
+    pub(super) fn execute_end_of_block_decorators(
+        &mut self,
+        basic_block_node: &BasicBlockNode,
+        node_id: MastNodeId,
+        current_forest: &Arc<MastForest>,
+        host: &mut impl AsyncHost,
+    ) -> ControlFlow<BreakReason> {
+        let num_ops = basic_block_node.num_operations() as usize;
+        for decorator in current_forest.decorators_for_op(node_id, num_ops) {
+            self.execute_decorator(decorator, host)?;
+        }
+
+        ControlFlow::Continue(())
+    }
+}
+
+// HELPERS
+// ===============================================================================================
+
+/// Given the current operation being executed within a basic block, returns the appropriate
+/// continuation to add to the continuation stack if execution is stopped at this point.
+fn get_continuation_from_operation(
+    basic_block_node: &BasicBlockNode,
+    op_batch: &OpBatch,
+    node_id: MastNodeId,
+    batch_index: usize,
+    op_idx_in_batch: usize,
+) -> Continuation {
+    let last_op_idx_in_batch = op_batch.ops().len() - 1;
+    let last_batch_idx_in_block = basic_block_node.num_op_batches() - 1;
+
+    if op_idx_in_batch < last_op_idx_in_batch {
```

Comment:
One thing that's not obvious: when this is called from `execute_op_batch` at line 272, `op_idx_in_batch` is the operation that just executed. So `op_idx_in_batch + 1` in the first branch is the next operation to execute, not the current one.

Worth adding a comment about that?